core/lib/node_framework_derive/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use syn::{parse_macro_input, DeriveInput};
+
+use crate::macro_impl::{MacroImpl, MacroKind};
+
+pub(crate) mod helpers;
+mod labels;
+mod macro_impl;
+
+/// Derive macro for the `FromContext` trait.
+/// Allows to automatically fetch all the resources and tasks from the context.
+///
+/// # Example
+///
+/// ```
+/// use zksync_node_framework_derive::FromContext;
+/// # mod service {
+/// #     pub struct ServiceContext<'a> {
+/// #         _marker: std::marker::PhantomData<&'a ()>,
+/// #     }
+/// #     impl<'a> ServiceContext<'a> {
+/// #         pub fn get_resource<T>(&mut self) -> Result<T, crate::WiringError> {
+/// #             unimplemented!()
+/// #         }
+/// #         pub fn get_resource_or_default<T>(&mut self) -> T {
+/// #             unimplemented!()
+/// #         }
+/// #     }
+/// # }
+/// # pub enum WiringError {
+/// #     ResourceLacking { _name: String },
+/// # }
+/// # pub trait FromContext: Sized {
+/// #     fn from_context(context: &mut service::ServiceContext<'_>) -> Result<Self, WiringError>;
+/// # }
+/// # struct MandatoryResource;
+/// # struct OptionalResource;
+/// # struct ResourceWithDefault;
+/// #[derive(FromContext)]
+/// # #[ctx(local)]
+/// struct MyWiringLayerInput {
+///     // The following field _must_ be present in the context.
+///     mandatory_resource: MandatoryResource,
```

Comment:
I think this approach can be tweaked as follows:

- Require that all fields for derived `FromContext` / `IntoContext` implementations implement the corresponding trait itself. I.e., make implementations just delegate to this impls similar to `serde` traits.
- Implement `FromContext` / `IntoContext` for the relevant types.

```rust
impl<T: Resource + Clone> FromContext for T {
    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
        context.get_resource::<T>()
    }
}

impl<T: FromContext> FromContext for Option<T> {
    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
        Ok(match T::from_context(context) {
            Ok(resource) => Some(resource),
            Err(WiringError::ResourceLacking { .. }) => None,
            Err(err) => return Err(err),
        })
    }
}

mpl<T: Resource> IntoContext for T {
    fn into_context(self, context: &mut ServiceContext<'_>) -> Result<(), WiringError> {
        context.insert_resource(self)
    }
}

impl<T: IntoContext> IntoContext for Option<T> {
    fn into_context(self, context: &mut ServiceContext<'_>) -> Result<(), WiringError> {
        if let Some(this) = self {
            this.into_context(context)
        } else {
            Ok(())
        }
    }
}

// Wrapper to distinguish tasks from resources.
#[derive(Debug)]
pub struct ContextTask<T>(pub T);

impl<T: Task> IntoContext for ContextTask<T> {
    fn into_context(self, context: &mut ServiceContext<'_>) -> Result<(), WiringError> {
        let Self(task) = self;
        context.add_task(task);
        Ok(())
    }
}
```

The only thing that it doesn't cover is the "or default" extractor for `FromContext`. It is possible to implement via a wrapper like this:

```rust
#[derive(Debug)]
pub struct OrDefault<T>(pub T);

impl<T: Resource + Clone + Default> FromContext for OrDefault<T> {
    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
        Ok(Self(context.get_resource_or_default::<T>()))
    }
}
```

...but I feel it would be unnecessarily verbose. If we assume that wrappers always have form `Wrapper<T>(pub T)`, then it's possible to introduce their support in derive macros like this:

```rust
#[derive(FromContext)]
struct MyContext {
    #[with = OrDefault]
    app_health: AppHealthCheckResource,
}

// expands to
impl FromContext for MyContext {
    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
        Ok(Self {
            app_health: <OrDefault<_> as FromContext>::from_context(context)?.0,
        })
    }
}
```

Alternatively, use `serde`-inspired functional extractors:

```rust
pub fn get_or_default<T>(context: &mut ServiceContext<'_>) -> Result<T, WiringError>
where
    T: Resource + Clone + Default,
{
    Ok(context.get_resource_or_default::<T>())
}

#[derive(FromContext)]
struct MyContext {
    #[with = get_or_default]
    // ^ accepts a path to a function w/ corresponding signature (may be generic, as in this case)
    app_health: AppHealthCheckResource,
}

// expands into
impl FromContext for MyContext {
    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
        Ok(Self {
            app_health: or_default(context)?,
        })
    }
}

```

I'm eager to do this in a follow-up PR to not block your efforts / to expedite, but IMO it would be more idiomatic / modular.

Reply:
Thanks for the thoughtful comment!
Funnily, I initially started with an approach close to [what you propose](https://github.com/matter-labs/zksync-era/pull/2330/commits/103139859ba09803175832d024c85370f41db90d), but then met the following problems:

First, we still will need wrapper types exposed to people `ContextTask`, in your case, and it feels unnecessary. I don't want the system to be designed around the procedural macro, I want it to be a mere tool to reduce boilerplate. And it leads to some clunkiness, where `Option<ContextTask<T>>` works, but `ContextTask<Option<T>>` in a compile error related to the trait implementations that may confuse inexperienced developers.

Then, we're building a somewhat complex infrastructure around these traits, allowing extensibility, but at the same time making it harder to understand. My point of view is that it's easy to slap either `task` or `resource` on fields, and it's clear what they do. Supporting `Option` _feels_ natural, IMHO. Same for `default` parameter. On the other hand, generic `with` support allows developer to do `#[with = my_custom_getter]`, which again will increase the complexity, because now you have to go and check what this function does, and it can in theory get very messy if two such modifiers will be applied and the behavior will become dependent on the order of fields in the structure.

What I wanted to achieve here is an explicitly restricted macro, that is easy to use, but _at the same time_ as soon as macro does not work for you anymore, you either redesign your code (getting and putting stuff from context shouldn't normally be complex), you just go and implement a trait manually.

I see the point regarding inconsistency, though, and we may discuss it further to try to find an optimal solution.

Reply:
tl;dr the problem domain feels very narrow to me, and I want to avoid overgeneralization just for the sake of overgeneralization.

Reply:
We discussed the design in DMs a bit, and I reworked the macro according to the results of the discussion.